geoips/pydantic/products.py

```diff
@@ -0,0 +1,231 @@
+    """output_formatters."""
+
+    pass
+
+
+class FilenameFormatterArgumentsModel(BaseModel):
+    """Validate FilenameFormatter arguments."""
+
+    pass
+
+
+class AlgorithmArgumentsModel(BaseModel):
+    """Validate Algorithm arguments."""
+
+    pass
+
+
+class InterpolatorArgumentsModel(BaseModel):
+    """Validate Interpolator arguments."""
+
+    pass
+
+
+class ReaderArgumentsModel(BaseModel):
+    """Validate Reader step arguments."""
+
+    model_config = ConfigDict(extra="forbid")
+    area_def: str = Field(None, description="Area definition identifier.")
+    chans: List[str] = Field(None, description="List of channels to process.")
+    metadata_only: bool = Field(False, description="Flag for metadata-only processing.")
+    self_register: bool = Field(False, description="Flag for self-registration.")
+
+
+class ProductStepDefinitionModel(BaseModel):
+    """Validate step definition : name, arguments."""
+
+    model_config = ConfigDict(extra="forbid")
+    type: str = Field(..., description="plugin type")
+    name: str = Field(..., description="plugin name")
+    arguments: Dict[str, Any] = Field(default_factory=dict, description="step args")
+
+    @model_validator(mode="before")
+    def _validate_plugin_name(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        """Validate plugin name."""
+        if not values:
+            raise ValueError("The plugin_name cannot be empty")
+        plugin_name = values.get("name", "").lower()
+        plugin_type = values.get("type", " ").lower()
+
+        valid_plugin_names = get_plugin_names(plugin_type)
+        if plugin_name not in valid_plugin_names:
+            raise ValueError(
+                f"{plugin_name} is invalid. \n\t"
+                f"Must be one of {sorted(valid_plugin_names)}"
+            )
+
+        return values
+
+    @model_validator(mode="before")
+    def _validate_plugin_arguments(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Validate & organize details for each step   .
+
+        Parameters
+        ----------
+        values : dict
+            A dictionary of plugin data. The key is plugin type, and
+            the value consists of plugin name & arguments
+
+        Returns
+        -------
+        values : dict
+            A validated and structured dictionary with the following fields:
+
+            - `type` : str
+                The type of the plugin.
+            - `name` : str
+                The name of the plugin.
+            - `arguments` : dict
+                The arguments associated with the plugin.
+
+        """
+        if not values:
+            raise ValueError("Empty : Missing step details")
+
+        # arguments = values.get("arguments", {})
+        # Delegate arguments validation to each plugin type argument class
+        plugin_type = values.get("type", " ").lower()
+        plugin_type_camel_case = "".join(
+            [word.capitalize() for word in plugin_type.split("_")]
```

Comment:
won't this result in snake case (AaaaBbbbCccc) instead of camel case (aaaBbbbCccc)?